File: Top-Interview-Questions-Leetcode/Easy-Collections/design/min-stack.cpp
#include <iostream>
#include <vector>
#include <utility>

class MinStack {
private:
    
    std::vector<std::pair<int, int> > container;
    
public:
    MinStack() {
        
    }
    
    void push(int val) {
        if (container.size() == 0) {
            container.push_back({val, val});
        }
        else {
            int mn = container.back().second;
            container.push_back({val, std::min(mn, val)});
        }
    }
    
    void pop() {
        container.pop_back();
    }
    
    int top() {
        return container.back().first;
    }
    
    int getMin() {
        return container.back().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Top-Interview-Questions-Leetcode/Easy-Collections/design/min-stack.py
class StackNode:
    def __init__(self, val=0, mn=0, next=None):
        self.val = val
        self.mn = mn
        self.next = next

class MinStack:

    def __init__(self):
        self.head = None

    def push(self, val: int) -> None:
        if self.head:
            newHead = StackNode(val, min(val, self.head.mn), self.head)
            self.head = newHead
        else:
            self.head = StackNode(val, val)

    def pop(self) -> None:
        self.head = self.head.next

    def top(self) -> int:
        return self.head.val

    def getMin(self) -> int:
        return self.head.mn


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()